Sources/Hummingbird/Router/Trie/Trie+resolve.swift

```diff
@@ -18,130 +18,197 @@ extension RouterTrie {
+        @inlinable
+        mutating func resolve() -> (value: Value, parameters: Parameters)? {
+            guard let component = pathComponents.first else {
+                guard let value = values[trie.nodes[0].valueIndex] else {
+                    return nil
+                }
+
+                return (value: value, parameters: parameters)
+            }
+
+            var nextNodeIndex = 1
+            guard let node = descend(
+                component: component,
+                nextPathComponentIndex: 1,
+                nextNodeIndex: &nextNodeIndex
+            ) else {
+                return nil
+            }
+
+            if let value = values[node.valueIndex] {
+                return (value: value, parameters: parameters)
+            } else {
+                return nil
+            }
+        }
+
+        @inlinable
+        mutating func descend(
+            component: Substring,
+            nextPathComponentIndex: Int,
+            nextNodeIndex: inout Int
+        ) -> TrieNode? {
+            var node = self.matchComponent(component, atNodeIndex: &nextNodeIndex)
+            var nextPathComponentIndex = nextPathComponentIndex
 
-        while let component = pathComponentsIterator.next() {
-            node = self.matchComponent(component, atNodeIndex: &nodeIndex, parameters: &parameters)
             if node.token == .recursiveWildcard {
                 // we have found a recursive wildcard. Go through all the path components until we match one of them
                 // or reach the end of the path component array
                 var range = component.startIndex..<component.endIndex
-                while let component = pathComponentsIterator.next() {
-                    var nodeIndex = nodeIndex
-                    let recursiveNode = self.matchComponent(component, atNodeIndex: &nodeIndex, parameters: &parameters)
+
+                while let component = nextPathComponent(advancingIndex: &nextPathComponentIndex) {
+                    var _nextNodeIndex = nextNodeIndex
+                    let recursiveNode = self.matchComponent(component, atNodeIndex: &_nextNodeIndex)
                     if recursiveNode.token != .deadEnd {
                         node = recursiveNode
+                        nextNodeIndex = _nextNodeIndex
                         break
                     }
                     // extend range of catch all text
                     range = range.lowerBound..<component.endIndex
                 }
                 parameters.setCatchAll(path[range])
             }
+
             if node.token == .deadEnd {
                 return nil
             }
-        }
 
-        if let value = self.values[node.valueIndex] {
-            return (value: value, parameters: parameters)
-        } else {
-            return nil
-        }
-    }
+            if let nextComponent = nextPathComponent(advancingIndex: &nextPathComponentIndex) {
+                // There's another component to the route
```

Comment:
We can do this because the trie nodes are ordered in priority?

Reply:
Correct.